### PR TITLE
WIP: Start hand-written PHP combinator-based parser

### PR DESCRIPTION
**Pull request by dmsnell**
**Foreward**

Clearly the PHP parser is the harder parser to write because JavaScript is much more amenable to the kind of programming style needed to effectively and efficiently write a parser. This PR is currently in progress experimenting with an idea which I believe will end up much more efficient than the generated parser from the PEG.

This code is still highly volatile and likely to change. That being said I welcome your feedback along the way. :smile:

I'm testing this code outside of WordPress. Locally I installed the PHPUnit `phar` and am running this to test…

```bash
php phpunit.phar --no-configuration --colors phpunit/class.block-parser-test.php
```

Please note that some cruft is lying around and there are style violations. Please know that I'm aware of these minutia and have no intention of leaving them in the final versions of this PR :wink:

---

The goal of this alternative parser is to make a faster parser than the
one generated by the PEG grammar directly. It must not take multiple
seconds to parse a document no matter how big.

This parser currently is designed to produce the blocks and their
attributes, leaving all inside content raw.

This is based on the seminal paper "Monadic Parser Combinators" by
Hutton and Meijer, University of Nottingham, 1996. I have had a hard
time finding literature detailing the performance of such an approach in
PHP.

Although this approach will be building up user-defined function
calls it avoids creating closures by means of passing around
descriptions of parsers in the PHP callback style. That is, instead of
passing partial functions, we pass a callable string and an array of
partial arguments. The input is added as the final input before calling
the given parser.

A further speedup taken in this approach involves defining parse rules "on
the branches" so as to reduce backtracking. That is, if two rules share
a prefix then we can parse the prefix and descend into a `first_of`
branch instead of having a top-level `first_of` branch and duplicating
the prefix. The latter approach has been taken in the formal grammar
spec to make the rules easier to read.

RegExp patterns have also been dupilcated where possible to combine
rules. That is, although we could ignore whitespace and then ignore a
closing block comment, instead I'm combining the RegExp pattern so that
we benefit from the faster performance of the RegExp and from only
making one call instead of two. Each call passes around its own state on
the parse "stack."

### CODE REVIEW

File: lib/block-parser.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+	const MAX_RUNTIME         = 1; // give up after one second
+
+	public function parse($input) {
+		$tic = microtime( true );
+
+		// trampoline for stack-safe recursion of the actual parser
+		while ( $this->input && ( microtime( true ) - $tic ) < self::MAX_RUNTIME ) {
+			return $this->proceed( $input );
+		}
```
Comment (nylen) [inline lib/block-parser.php]: I am not sure about the non-deterministic behavior introduced by the time limit here.  It will likely mean, for example, that some posts parse correctly on one page load but not on the next.  This is a bit worrisome to me, and it would be better to make sure we understand the performance characteristics of our parser(s) well.

File: lib/block-parser.php [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+		list( /* production */, $remaining ) = $result;
+
+		return array( array(), $remaining );
+	}
+
+	public static function literal( $value, $input ) {
+		return strpos( $input, $value ) === 0
+			? array( $value, substr( $input, strlen( $value ) ) )
```
Comment (nylen) [inline lib/block-parser.php]: Here and below, should we be using `mb_substr` instead?  If not, how can we guarantee that this parser will behave well with UTF-8 input?

If we do switch to `mb_substr`, then performance becomes a concern because it has to scan through the entire string until each start position.

File: lib/block-parser.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
+<?php
+
+if (!class_exists('Gutenberg_Block_Parser_State', false)):
```
Comment (nylen) [inline lib/block-parser.php]: Why would these classes already exist?

**Review by nylen — COMMENTED** [review]
In general I am finding this code a bit hard to follow.  That is probably because I am not well-versed in the theory at play here.  However, this is a general concern about our parsing implementation:  so far, it seems like only a few people will have the knowledge needed to maintain these.

There are several categories of tests that are absolutely mandatory that we add very soon, especially before switching over to a parsing solution that is different than "use the same grammar on the client and server":

- Invalid nesting of blocks, malformed delimiters, content that cuts off before the end of a block delimiter, etc.
- Multibyte character handling, including block delimiters that are made invalid by the insertion of multibyte characters in various places

Finally, I question whether the assumptions that led to the creation of this PR are actually valid.  It seems possible to automatically generate a quite performant parser based on a PEG.js grammar.  See #1775 for details.

We need to do similar profiling of this solution.  In particular there will be thousands of `call_user_func` calls every time a post is parsed, and we can likely save a lot of time by getting rid of these.

File: phpunit/class.block-parser-test.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		return $this->parser->parse( $input );
+	}
+
+	function setUp() {
+		$this->parser = new Gutenberg_Block_Parser();
+	}
+
+	function test_combinator_succeed() {
+		$this->assertEquals(
+			[ [ 'test', 'bork' ] ],
```
Comment (dmsnell) [inline phpunit/class.block-parser-test.php]: it's only in the test code

File: lib/block-parser.php (hunk 3 above)
Comment (dmsnell) [inline lib/block-parser.php]: hopefully they wouldn't. just guarding

File: lib/block-parser.php (hunk 1 above)
Comment (dmsnell) [inline lib/block-parser.php]: well clearly it's not fully built out. there will be some post that's about two million lines long and that will hang on the server. this isn't a big part of this PR but the general idea is that it provides a safe limit where if we run out of time can stop parsing and return the rest as unparsed

again, not a big part of this (unfinished) PR

File: lib/block-parser.php (hunk 2 above)
Comment (dmsnell) [inline lib/block-parser.php]: it's a good point and I was planning on creating a local string length function. one thing I wanted to check on is if _not_ using `mb_substr` is actually going to cause problems, since our syntax is all ASCII. we shouldn't theoretically be splitting on any non-ASCII characters and so I think we can probably make this assumption safely (though I haven't finished considering that yet)

File: lib/block-parser.php [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+endif;
+
+if (!class_exists('Gutenberg_Block_Parser', false)):
+
+class Gutenberg_Block_Parser {
+	const BLOCK_COMMENT_OPEN  = '(^<!--)';
+	const BLOCK_COMMENT_CLOSE = '(^/?-->)';
+	const BLOCK_NAME          = '(^[[:alpha:]](?:[[:alnum:]]|/[[:alnum:]])*)i';
```
Comment (nb) [inline lib/block-parser.php]: Can we list this to the registered blocks and avoid regular expressions altogether?

File: phpunit/class.block-parser-test.php (hunk 4 above)
Comment (nylen) [inline phpunit/class.block-parser-test.php]: Yes, which will need to me merged into the plugin and run against PHP 5.2 along with the rest of the tests.

File: lib/block-parser.php (hunk 5 above)
Comment (nylen) [inline lib/block-parser.php]: We still need to be able to safely parse blocks from plugins that have been disabled or uninstalled. This could work by including them within larger chunks of content, but I think it is better to show them as an independent but "unrecognized" block.  See #1735.

I also don't like the idea of coupling the parser to the list of registered blocks.

File: lib/block-parser.php (hunk 2 above)
Comment (nylen) [inline lib/block-parser.php]: It's indeed possible that this won't be an issue here. Let's test it thoroughly rather than assuming.

File: lib/block-parser.php (hunk 1 above)
Comment (nylen) [inline lib/block-parser.php]: 👍  Generally I would hope that we can do _better_ than the rest of WP core here.

File: lib/block-parser.php (hunk 5 above)
Comment (dmsnell) [inline lib/block-parser.php]: @nb I don't think that we should totally avoid RegExps. In this case I'm only using simple RegExps to tokenize and not to parse. As @nylen pointed out we unfortunately can't limit this to known block types because it has to parse everything.

The grammar doesn't use RegExps and we wouldn't have to here, but I felt like it was going to be faster and clearer than running character-by-character the way the PEG parser does.

**Comment by dmsnell** [comment]
I'm planning on closing this but might play a little more with the concept. There's no way to actually test the performance until the full grammar is incorporated, so maybe I'll try and fill that out and test.

**Comment by dmsnell** [comment]
Okay so my curiosity got to me and I don't think I'll close this yet. Right now it handles most of the grammar and I'd really like to see if how its performance compares to the generated parser. I wouldn't be at all surprised if it's much slower actually.

It's still a WIP so don't be surprised at the ugly `proceed()` function. There are things I don't like about it but yet that is still the place where we diverge from the pattern of monadic parsers and start thinking iteratively with a stack and state.

Some of the returned values from the parsers I have been thinking of changing, plus this is still all immutable. These changes could speed things up dramatically especially for long documents. In the first version I wanted to make sure I could do it right and immutability has helped a lot by keeping different things different.

One big point that I think I need to add in the values from the parsers is some indication of which production is being generated. It's hard right now to use the results of the `first_of` combinator because you have to structurally determine which of the productions you have; this is impossible if the productions contain the same number of values.

If this all works well we might be able to turn the `proceed()` function into a parser too. I think it would be something like…

```
block => sequence [
  block_opener,
  zero_or_more [
    block,
    raw_content
  ],
  block_closer
]
```

…but the one big problem there is that if this isn't fast enough we have no safe exit. with `process()` imperatively built we can set a threshold on the parse and when we reach the limit, return the remainder of the document as a freeform block - a concept I find pretty cool. this would obviously depend on Gutenberg not killing blocks it doesn't understand though

**Comment by dmsnell** [comment]
> In general I am finding this code a bit hard to follow. That is probably because I am not well-versed in the theory at play here. However, this is a general concern about our parsing implementation: so far, it seems like only a few people will have the knowledge needed to maintain these.

This is something I also recognize but at the same time I feel like if we want to accomplish our goals we will have to accept that the theory is there to enable us and will keep us safe and successful even if it _is_ unfamiliar.

As it becomes more stable I would expect more of our efforts to go into documenting the concepts and building guides so that more people can learn about the theory and how to participate in this code (all parsing code). It will help the project and hopefully help them too.

That raises a point I'd like to look into and that is making the parser less integral to Gutenberg (as you did similar work earlier) and let it be overwritable by a plugin. We have to cover both the parse and the serialize functions. In my head I would like to explore some other approaches that don't involve serializing to a string…

 - save JSON blob directly to post meta
 - save structure to Simperium as-is
 - API calls instead of local processing
 - something else fun, maybe directly into a (gasp, MongoDB) database table?

These approaches don't actually need string parsing (other than maybe the native JSON parse) and could offer novel ways to speed up the editor for heavy workloads or integrate into other storage systems.

Anyway, still doodling here

**Comment by dmsnell** [comment]
It's not necessary to keep this open so I'm going to close it. The idea _may_ be worth exploring more but I don't have the time now to do it.
